File: 02-tool-system/src/loop.ts
/**
 * Agent 对话循环（第二版：集成工具注册表和权限管线）
 * 对照 Claude Code 源码 src/query.ts:
 *   - queryLoop (L241-1729): 1500 行
 *   - while(true) (L307): 相同循环结构
 *   - L1382: tool_use 处理入口 → toolOrchestration
 */

import type { Message, ContentBlock, LoopState, Terminal, StreamEvent, Tool } from "./types.js";
import { ToolExecutor } from "./executor.js";

export async function* agentLoop(
  state: LoopState,
  callModel: (messages: Message[], tools: Tool[]) => Promise<ContentBlock[]>,
  tools: Tool[],
  executor: ToolExecutor,
  options: { maxTurns?: number } = {},
): AsyncGenerator<StreamEvent, Terminal> {
  const maxTurns = options.maxTurns ?? 10;

  while (true) {
    const { messages, turnCount } = state;

    if (turnCount >= maxTurns) {
      return { reason: "max_turns" };
    }

    // 1. 调用模型
    const assistantContent = await callModel(messages, tools);
    for (const block of assistantContent) {
      if (block.type === "text") yield { type: "text_delta", text: block.text };
      else if (block.type === "tool_use") yield { type: "tool_use", id: block.id, name: block.name, input: block.input };
    }

    // 2. 收集 tool_use
    const toolCalls = assistantContent.filter(
      (b): b is Extract<ContentBlock, { type: "tool_use" }> => b.type === "tool_use",
    );

    if (toolCalls.length === 0) {
      yield { type: "turn_complete", turn: turnCount };
      return { reason: "completed" };
    }

    // 3. 通过执行器执行工具（内含权限检查）
    const { results: toolResults, events } = await executor.executeCalls(toolCalls);
    for (const evt of events) {
      yield evt;
    }

    yield { type: "turn_complete", turn: turnCount };

    // 4. 回填结果，进入下一轮
    state = {
      messages: [
        ...messages,
        { role: "assistant", content: assistantContent },
        { role: "user", content: toolResults },
      ],
      turnCount: turnCount + 1,
    };
  }
}

File: 06-rust-rewrite/src/loop_.rs
/// Agent 循环
///
/// 对照 TS 版 loop.ts：async function* + while(true) + yield
/// Rust 版用 mpsc channel 发事件，效果一样但类型更安全

use crate::model::ModelCaller;
use crate::permission::PermissionGate;
use crate::tools::ToolRegistry;
use crate::types::{
    ContentBlock, LoopState, PermissionDecision, StreamEvent, TokenBudget,
};

pub async fn agent_loop(
    state: &mut LoopState,
    caller: &dyn ModelCaller,
    registry: &ToolRegistry,
    gate: &PermissionGate,
    budget: &TokenBudget,
    max_turns: usize,
    tx: &tokio::sync::mpsc::Sender<StreamEvent>,
) -> Result<String, String> {
    loop {
        if state.turn_count >= max_turns {
            tx.send(StreamEvent::LoopComplete {
                reason: "max_turns".into(),
            })
            .await
            .map_err(|e| e.to_string())?;
            return Ok("max_turns".into());
        }

        // Token 预算检查
        let used = TokenBudget::estimate(&state.messages);
        tx.send(StreamEvent::TokenBudget {
            used,
            max: budget.max_tokens,
            threshold: budget.compact_threshold,
        })
        .await
        .map_err(|e| e.to_string())?;

        // 1. 调用模型——对照 TS 版 callModel(messages)
        let blocks = caller.call(&state.messages).await;

        // 2. 发送文本和工具调用事件
        let mut tool_calls: Vec<ContentBlock> = Vec::new();
        for block in &blocks {
            match block {
                ContentBlock::Text { text } => {
                    tx.send(StreamEvent::TextDelta { text: text.clone() })
                        .await
                        .map_err(|e| e.to_string())?;
                }
                ContentBlock::ToolUse { id, name, input } => {
                    tx.send(StreamEvent::ToolUse {
                        id: id.clone(),
                        name: name.clone(),
                        input: input.clone(),
                    })
                    .await
                    .map_err(|e| e.to_string())?;
                    tool_calls.push(block.clone());
                }
                _ => {}
            }
        }

        // 3. 没有工具调用 → 结束
        if tool_calls.is_empty() {
            tx.send(StreamEvent::TurnComplete {
                turn: state.turn_count,
            })
            .await
            .map_err(|e| e.to_string())?;
            tx.send(StreamEvent::LoopComplete {
                reason: "completed".into(),
            })
            .await
            .map_err(|e| e.to_string())?;
            return Ok("completed".into());
        }

        // 4. 执行工具
        let mut results: Vec<ContentBlock> = Vec::new();
        for call in &tool_calls {
            if let ContentBlock::ToolUse { id, name, input } = call {
                if let Some(tool) = registry.get(name) {
                    // 权限检查
                    let perm = gate.check(tool.as_ref(), input);
                    let decision = perm.decision;
                    tx.send(StreamEvent::PermissionCheck {
                        tool: name.clone(),
                        decision,
                    })
                    .await
                    .map_err(|e| e.to_string())?;

                    match decision {
                        PermissionDecision::Deny => {
                            results.push(ContentBlock::ToolResult {
                                tool_use_id: id.clone(),
                                content: format!("权限拒绝: {}", perm.reason),
                                is_error: true,
                            });
                        }
                        _ => {
                            // 执行
                            let result = tool.execute(input.clone()).await;
                            match result {
                                Ok(content) => {
                                    results.push(ContentBlock::ToolResult {
                                        tool_use_id: id.clone(),
                                        content,
                                        is_error: false,
                                    });
                                }
                                Err(e) => {
                                    results.push(ContentBlock::ToolResult {
                                        tool_use_id: id.clone(),
                                        content: format!("错误: {}", e),
                                        is_error: true,
                                    });
                                }
                            }
                        }
                    }
                } else {
                    results.push(ContentBlock::ToolResult {
                        tool_use_id: id.clone(),
                        content: format!("未知工具: {}", name),
                        is_error: true,
                    });
                }
            }
        }

        // 发送工具结果事件
        for r in &results {
            if let ContentBlock::ToolResult {
                tool_use_id, content, ..
            } = r
            {
                tx.send(StreamEvent::ToolResult {
                    tool_use_id: tool_use_id.clone(),
                    content: content.clone(),
                })
                .await
                .map_err(|e| e.to_string())?;
            }
        }

        tx.send(StreamEvent::TurnComplete {
            turn: state.turn_count,
        })
        .await
        .map_err(|e| e.to_string())?;

        // 更新状态
        state.messages.push(crate::types::ChatMessage {
            role: "assistant".into(),
            content: blocks,
        });
        state.messages.push(crate::types::ChatMessage {
            role: "user".into(),
            content: results,
        });
        state.turn_count += 1;
    }
}
